File: src/main/services/document-store.ts
import path from 'node:path';

import type { AnalysisResult, DocumentItem, ItemStatus } from '../../../shared/types.js';

type StoredDocument = Omit<DocumentItem, 'displayName'>;

export class DocumentStore {
  private readonly items = new Map<string, StoredDocument>();

  list(): DocumentItem[] {
    return [...this.items.values()].map((item) => this.toDocumentItem(item));
  }

  add(paths: string[]): DocumentItem[] {
    for (const resolvedPath of paths) {
      if (this.items.has(resolvedPath)) {
        continue;
      }
      this.items.set(resolvedPath, {
        key: resolvedPath,
        sourcePath: resolvedPath,
        currentPath: resolvedPath,
        status: 'pending',
        analysis: null,
        proposedName: '',
        errorMessage: '',
        skipped: false,
        history: [],
      });
    }
    return this.list();
  }

  clear(): DocumentItem[] {
    this.items.clear();
    return [];
  }

  get(key: string): DocumentItem | null {
    const item = this.items.get(key);
    return item ? this.toDocumentItem(item) : null;
  }

  getMany(keys?: string[]): DocumentItem[] {
    if (!keys || keys.length === 0) {
      return this.list();
    }
    return keys
      .map((key) => this.get(key))
      .filter((item): item is DocumentItem => Boolean(item));
  }

  updateAnalysis(key: string, analysis: AnalysisResult, proposedName: string): DocumentItem {
    const current = this.requireItem(key);
    current.analysis = analysis;
    current.proposedName = proposedName;
    current.errorMessage = '';
    current.status = analysis.confidence >= 0.8 ? 'ready' : 'needs_review';
    current.skipped = false;
    current.history = [...current.history, `analyzed:${JSON.stringify(analysis)}`];
    return this.toDocumentItem(current);
  }

  markAnalyzing(key: string): DocumentItem {
    return this.patch(key, {
      status: 'analyzing',
      errorMessage: '',
      skipped: false,
    });
  }

  markError(key: string, message: string): DocumentItem {
    const current = this.requireItem(key);
    current.status = 'error';
    current.errorMessage = message;
    current.history = [...current.history, `error:${message}`];
    return this.toDocumentItem(current);
  }

  markSkipped(key: string): DocumentItem {
    const current = this.requireItem(key);
    current.status = 'skipped';
    current.skipped = true;
    current.history = [...current.history, 'skipped'];
    return this.toDocumentItem(current);
  }

  markRenamed(key: string, nextPath: string): DocumentItem {
    const current = this.requireItem(key);
    current.currentPath = nextPath;
    current.status = 'renamed';
    current.history = [...current.history, `renamed:${path.basename(nextPath)}`];
    return this.toDocumentItem(current);
  }

  updateProposedName(key: string, proposedName: string): DocumentItem {
    return this.patch(key, { proposedName });
  }

  setStatus(key: string, status: ItemStatus): DocumentItem {
    return this.patch(key, { status });
  }

  private patch(key: string, partial: Partial<StoredDocument>): DocumentItem {
    const current = this.requireItem(key);
    const next = { ...current, ...partial };
    this.items.set(key, next);
    return this.toDocumentItem(next);
  }

  private requireItem(key: string): StoredDocument {
    const item = this.items.get(key);
    if (!item) {
      throw new Error(`対象ドキュメントが見つかりません: ${key}`);
    }
    return item;
  }

  private toDocumentItem(item: StoredDocument): DocumentItem {
    return {
      ...item,
      displayName: path.basename(item.currentPath),
    };
  }
}

File: src-tauri/src/services/document_store.rs
use std::path::Path;

use crate::types::{AnalysisResult, DocumentItem, ItemStatus};

pub struct DocumentStore {
    items: indexmap::IndexMap<String, DocumentItem>,
}

impl DocumentStore {
    pub fn new() -> Self {
        Self {
            items: indexmap::IndexMap::new(),
        }
    }

    pub fn list(&self) -> Vec<DocumentItem> {
        self.items.values().cloned().collect()
    }

    pub fn add(&mut self, paths: Vec<String>) -> Vec<DocumentItem> {
        for path in paths {
            if self.items.contains_key(&path) {
                continue;
            }
            let display_name = Path::new(&path)
                .file_name()
                .map(|n| n.to_string_lossy().to_string())
                .unwrap_or_else(|| path.clone());
            self.items.insert(
                path.clone(),
                DocumentItem {
                    key: path.clone(),
                    source_path: path.clone(),
                    current_path: path,
                    display_name,
                    status: ItemStatus::Pending,
                    analysis: None,
                    proposed_name: String::new(),
                    error_message: String::new(),
                    skipped: false,
                    history: vec![],
                },
            );
        }
        self.list()
    }

    pub fn clear(&mut self) -> Vec<DocumentItem> {
        self.items.clear();
        vec![]
    }

    pub fn get(&self, key: &str) -> Option<DocumentItem> {
        self.items.get(key).cloned()
    }

    pub fn get_many(&self, keys: Option<&[String]>) -> Vec<DocumentItem> {
        match keys {
            None => self.list(),
            Some(ks) if ks.is_empty() => self.list(),
            Some(ks) => ks.iter().filter_map(|k| self.get(k)).collect(),
        }
    }

    pub fn update_analysis(
        &mut self,
        key: &str,
        analysis: AnalysisResult,
        proposed_name: String,
    ) -> Option<DocumentItem> {
        let item = self.items.get_mut(key)?;
        item.status = if analysis.confidence >= 0.8 {
            ItemStatus::Ready
        } else {
            ItemStatus::NeedsReview
        };
        item.error_message.clear();
        item.skipped = false;
        item.history
            .push(format!("analyzed:{}", serde_json::to_string(&analysis).unwrap_or_default()));
        item.analysis = Some(analysis);
        item.proposed_name = proposed_name;
        Some(item.clone())
    }

    pub fn mark_analyzing(&mut self, key: &str) -> Option<DocumentItem> {
        let item = self.items.get_mut(key)?;
        item.status = ItemStatus::Analyzing;
        item.error_message.clear();
        item.skipped = false;
        Some(item.clone())
    }

    pub fn mark_error(&mut self, key: &str, message: String) -> Option<DocumentItem> {
        let item = self.items.get_mut(key)?;
        item.status = ItemStatus::Error;
        item.error_message = message.clone();
        item.history.push(format!("error:{}", message));
        Some(item.clone())
    }

    pub fn mark_skipped(&mut self, key: &str) -> Option<DocumentItem> {
        let item = self.items.get_mut(key)?;
        item.status = ItemStatus::Skipped;
        item.skipped = true;
        item.history.push("skipped".to_string());
        Some(item.clone())
    }

    pub fn mark_renamed(&mut self, key: &str, next_path: String) -> Option<DocumentItem> {
        let item = self.items.get_mut(key)?;
        let base = Path::new(&next_path)
            .file_name()
            .map(|n| n.to_string_lossy().to_string())
            .unwrap_or_else(|| next_path.clone());
        item.history.push(format!("renamed:{}", base));
        item.current_path = next_path.clone();
        item.display_name = base;
        item.status = ItemStatus::Renamed;
        Some(item.clone())
    }

    pub fn update_proposed_name(&mut self, key: &str, proposed_name: String) -> Option<DocumentItem> {
        let item = self.items.get_mut(key)?;
        item.proposed_name = proposed_name;
        Some(item.clone())
    }
}
